src/platform/Darwin/DnssdImpl.cpp

```diff
@@ -219,25 +291,52 @@ static void OnGetAddrInfo(DNSServiceRef sdRef, DNSServiceFlags flags, uint32_t i
     ReturnOnFailure(MdnsContexts::GetInstance().Has(sdCtx));
     LogOnFailure(__func__, err);
 
+    std::string domainName = GetDomainFromHostName(hostname);
+    if (domainName.empty())
+    {
+        ChipLogError(Discovery, "Mdns: Domain name is not set in hostname %s", hostname);
+        return;
+    }
     if (kDNSServiceErr_NoError == err)
     {
-        sdCtx->OnNewAddress(interfaceId, address);
+        std::pair<uint32_t, std::string> key = std::make_pair(interfaceId, domainName);
+        sdCtx->OnNewAddress(key, address);
     }
 
     if (!(flags & kDNSServiceFlagsMoreComing))
```

Comment:
Can `kDNSServiceFlagsMoreComing` be relied upon to work this way when `kDNSServiceFlagsShareConnection` is in use? It seems like we're assuming here that it means the `DNSServiceBrowse` and `DNSServiceGetAddrInfo` calls for the domain of this last result is done for now.

https://developer.apple.com/documentation/dnssd/1823436-anonymous/kdnsserviceflagsshareconnection

Reply:
When Boris, Ted and I discussed we didn't realize this can be an issue. I need to spend some time on this as I am not an expert in the domain. @bzbarsky-apple do you have an opinion here?

Reply:
kDNSServiceFlagsMoreComing says that we have any more data on the shared connection. If you are doing multiple queries on the same connection, the response that doesn't set the MoreComing flag for any particular series of results can happen on any of the queries.

Reply:
This needs more thought and has to be fixed correctly. I need to check the API documentation and see how best to handle the kDNSServiceFlagsMoreComing for a shared connection. will do as a follow up

Reply:
filed - #32681